pkg/testing/tools/artifacts_snapshot.go

```diff
@@ -0,0 +1,204 @@
+	manResponse, err := c.getManifestResponse(ctx, version)
+	if err != nil {
+		return fmt.Errorf("getting manifest response: %w", err)
+	}
+	manifestPackages, err := c.getManifestPackages(ctx, manResponse.ManifestURL)
+	if err != nil {
+		return fmt.Errorf("getting manifest packages: %w", err)
+	}
+	for _, pkg := range packageRequests {
+		pkgData, ok := manifestPackages[pkg.Name]
+		if !ok {
+			return fmt.Errorf("package %s not found in manifest", pkg)
+		}
+		err := c.downloadPackage(ctx, pkgData, pkg.TargetPath)
+		if err != nil {
+			return fmt.Errorf("downloading package: %w", err)
+		}
+	}
+	return nil
+}
+
+func (c *ArtifactSnapshotClient) getManifestResponse(ctx context.Context, version string) (*ArtifactsManifestResponse, error) {
+	endpoint := fmt.Sprintf(beatsManifestEndpoint, version)
+	url := fmt.Sprintf("%s/%s", c.url, endpoint)
+	response, err := c.createAndPerformRequest(ctx, url)
+	if err != nil {
+		return nil, fmt.Errorf("creating and performing request: %w", err)
+	}
+	manResponse, err := checkResponseAndUnmarshal[ArtifactsManifestResponse](response)
+	if err != nil {
+		return nil, fmt.Errorf("checking response and unmarshalling: %w", err)
+	}
+	return manResponse, nil
+}
+
+func (c *ArtifactSnapshotClient) getManifestPackages(ctx context.Context, manifestURL string) (map[string]PackageData, error) {
+	response, err := c.createAndPerformRequest(ctx, manifestURL)
+	if err != nil {
+		return nil, fmt.Errorf("creating and performing request: %w", err)
+	}
+	manResponse, err := checkResponseAndUnmarshal[Projects](response)
+	if err != nil {
+		return nil, fmt.Errorf("checking response and unmarshalling: %w", err)
+	}
+	allPackages := make(map[string]PackageData)
+	for _, v := range maps.Values(manResponse.Projects) {
+		for k, pack := range v.Packages {
+			allPackages[k] = pack
+		}
+	}
+	return allPackages, nil
+}
+
+func (c *ArtifactSnapshotClient) downloadPackage(ctx context.Context, pkg PackageData, destinationDir string) error {
+
+	return nil
+}
+
+func downloadFile(url string, path string) error {
+	var filePath string
+	if path == "" {
+		tempParentDir := filepath.Join(os.TempDir(), uuid.NewString())
+		filePath = filepath.Join(tempParentDir, uuid.NewString())
+		path = filePath
+	} else {
+		filePath = filepath.Join(path, uuid.NewString())
+	}
+
+	tempFile, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	defer tempFile.Close()
+	exp := getExponentialBackOff(3)
+
+	retryCount := 1
+	var fileReader io.ReadCloser
+	download := func() error {
+		resp, err := http.Get(url)
+		if err != nil {
+			retryCount++
+			return fmt.Errorf("getting url %s: %w", url, err)
+		}
+		fileReader = resp.Body
+
+		return nil
+	}
+
+	err = backoff.Retry(download, exp)
+	if err != nil {
+		return err
+	}
+	defer fileReader.Close()
+
+	_, err = io.Copy(tempFile, fileReader)
+	if err != nil {
+		return fmt.Errorf("copying file: %w", err)
+	}
+
+	_ = os.Chmod(tempFile.Name(), 0666)
```

Comment:
[Blocker]
Handle the error. If this fails, it might lead to unexpected behaviour which will be hard to understand as we'd missing this error